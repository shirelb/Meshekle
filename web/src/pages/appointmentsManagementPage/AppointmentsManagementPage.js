import React from 'react';
import './styles.css';

import moment from 'moment';

import {Grid, Header} from 'semantic-ui-react';
import 'semantic-ui-css/semantic.min.css';
import AppointmentCalendar from "../../components/calendars/AppointmentCalendar";
import store from 'store';
import {Helmet} from 'react-helmet';
import strings from "../../shared/strings";
import AppointmentInfo from "../../components/appointment/AppointmentInfo";
import AppointmentEdit from "../../components/appointment/AppointmentEdit";
import {Route, Switch} from "react-router-dom";
import AppointmentAdd from "../../components/appointment/AppointmentAdd";
import AppointmentRequestInfo from "../../components/appointmentRequest/AppointmentRequestInfo";
import DraggableAppointmentRequest from "../../components/appointmentRequest/DraggableAppointmentRequest";
import appointmentsStorage from "../../storage/appointmentsStorage";
import usersStorage from "../../storage/usersStorage";

const TOTAL_PER_PAGE = 10;


class AppointmentsManagementPage extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            appointments: [],
            appointmentRequests: [],
            calendarEvents: [],
            page: 0,
            totalPages: 0,
            openPopup: false,
            eventPopup: {},
            highlightTableRow: null,
            appointmentRequestHoovering: {requestId: -1},
        };

        this.incrementPage = this.incrementPage.bind(this);
        this.decrementPage = this.decrementPage.bind(this);
        this.setPage = this.setPage.bind(this);
        this.handleDelete = this.handleDelete.bind(this);
        this.updateAfterMoveOrResizeEvent = this.updateAfterMoveOrResizeEvent.bind(this);

        this.serviceProviderHeaders = '';
    }

    componentDidMount() {
        this.serviceProviderHeaders = {
            'Authorization': 'Bearer ' + store.get('serviceProviderToken')
        };
        this.userId = store.get('userId');
        this.serviceProviderId = store.get('serviceProviderId');
        this.getServiceProviderAppointments();
        this.getServiceProviderAppointmentRequests();
    }

    getServiceProviderAppointmentRequests() {
        appointmentsStorage.getServiceProviderAppointmentRequests(this.serviceProviderId, this.serviceProviderHeaders)
            .then((response) => {
                const appointmentRequests = response.data;
                const totalPages = Math.ceil(appointmentRequests.length / TOTAL_PER_PAGE);

                appointmentRequests.map((appointmentRequest, index) => {
                    usersStorage.getUserByUserID(appointmentRequest.AppointmentDetail.clientId, this.serviceProviderHeaders)
                        .then(user => {
                            appointmentRequest.clientName = user.fullname;
                            appointmentRequest.optionalTimes = JSON.parse(appointmentRequest.optionalTimes);

                            let appointmentRequestEvent = {
                                id: appointmentRequest.requestId,
                                title: appointmentRequest.clientName,
                                allDay: false,
                                start: null,
                                end: null,
                                appointmentRequest: appointmentRequest,
                                // color:'#b7d2ff',
                                backgroundColor: '#45b0d9',
                            };

                            let appointmentRequestsEvents = this.state.appointmentRequests;
                            appointmentRequestsEvents.push(appointmentRequestEvent);
                            this.setState({
                                appointmentRequests: appointmentRequestsEvents
                            });
                        });
                });

                this.setState({
                    page: 0,
                    totalPages,
                });
            });
    }


    componentWillReceiveProps({location = {}}) {
        if (location.pathname === '/appointments' && location.pathname !== this.props.location.pathname) {
            this.setState({
                appointments: [],
                appointmentRequests: [],
            });
            this.getServiceProviderAppointments();
            this.getServiceProviderAppointmentRequests();
        }
    }

    getServiceProviderAppointments() {
        appointmentsStorage.getServiceProviderAppointments(this.serviceProviderId, this.serviceProviderHeaders)
            .then((response) => {
                this.setState({appointments: []});

                const appointments = response.data;
                const totalPages = Math.ceil(appointments.length / TOTAL_PER_PAGE);

                appointments.map((appointment, index) => {
                    usersStorage.getUserByUserID(appointment.AppointmentDetail.clientId, this.serviceProviderHeaders)
                        .then(user => {
                            let appointmentEvent = {};
                            appointment.clientName = user.fullname;

                            appointmentEvent.id = appointment.appointmentId;
                            appointmentEvent.title = user.fullname;
                            appointmentEvent.allDay = false;
                            appointmentEvent.start = moment(appointment.startDateAndTime);
                            appointmentEvent.end = moment(appointment.endDateAndTime);
                            appointmentEvent.appointment = appointment;

                            let appointmentsEvents = this.state.appointments;
                            appointmentsEvents.push(appointmentEvent);
                            this.setState({
                                appointments: appointmentsEvents
                            });
                        });
                });

                this.setState({
                    page: 0,
                    totalPages,
                });

            });
    }

    setPage(page) {
        return () => {
            this.setState({page});
        };
    }

    decrementPage() {
        const {page} = this.state;

        this.setState({page: page - 1});
    }

    incrementPage() {
        const {page} = this.state;

        this.setState({page: page + 1});
    }

    handleDelete(appointmentId) {
        const {appointments} = this.state;

        this.setState({
            appointments: appointments.filter(a => a.id !== appointmentId),
        });
    }

    hoverOnAppointmentRequest = (appointmentRequest) => {
        let isAppointmentsWithOptional = this.state.appointments.filter(obj => obj.id === appointmentRequest.requestId && obj.status === 'optional');
        if (isAppointmentsWithOptional.length > 0)
            return;

        this.setState({appointmentRequestHoovering: appointmentRequest});
        let shadowAppointments = [];
        if (Array.isArray(appointmentRequest.optionalTimes)) {
            appointmentRequest.optionalTimes.map(datesTimes => {
                datesTimes.hours.map(time => {
                    shadowAppointments.push(
                        {
                            id: appointmentRequest.requestId,
                            title: appointmentRequest.clientName,
                            allDay: false,
                            start: moment(moment(datesTimes.date).format("YYYY-MM-DD") + ' ' + time.startHour).toDate(),
                            end: moment(moment(datesTimes.date).format("YYYY-MM-DD") + ' ' + time.endHour).toDate(),
                            appointmentRequest: appointmentRequest,
                            status: "optional",
                            // color:'#b7d2ff',
                            backgroundColor: '#45b0d9',
                        }
                    )
                })
            });
            shadowAppointments.push.apply(shadowAppointments, this.state.appointments);
            if (Array.isArray(shadowAppointments)) {
                this.setState({appointments: shadowAppointments});
            }
        }
    };

    hoverOffAppointmentRequest = (appointmentRequest) => {
        let appointmentsWithoutOptional = this.state.appointments.filter(obj => obj.id !== appointmentRequest.requestId || obj.status !== 'optional');
        this.setState({appointments: appointmentsWithoutOptional});
    };

    onSelectEvent = event => {
        console.log('onSelectEvent=event  ', event);
        // if (event.appointmentId === this.state.highlightTableRow)
        //     this.setState({highlightTableRow: null});
        // else {
        //     this.setState({highlightTableRow: event.appointmentId});
        //
        // }
        this.props.history.push(`${this.props.match.path}/${event.appointment.appointmentId}`, {
            appointment: event.appointment
        });
    };

    onSelectSlot = (start, end) => {
        let slotInfo = {start: start, end: end};

        this.props.history.push(`${this.props.match.path}/set`, {
            slotInfo: slotInfo,
        });
    };

    approveAppointmentRequest = (appointmentRequestEventDropped) => {
        var appointmentRequests = this.state.appointmentRequests;
        appointmentsStorage.approveAppointmentRequestById(appointmentRequestEventDropped.appointmentRequest.requestId, this.serviceProviderHeaders)
            .then(response => {
                console.log('appointmentRequest approved ', response);

                let updatedAppointmentsRequests = appointmentRequests.filter(function (obj) {
                    return obj.appointmentRequest.requestId !== appointmentRequestEventDropped.appointmentRequest.requestId;
                });
                this.setState({appointmentRequests: updatedAppointmentsRequests})
            });
    };

    onDropAppointmentRequest = (appointmentRequestEvent) => {
        let appointmentRequestDropped = {
            id: appointmentRequestEvent.id,
            title: appointmentRequestEvent.title,
            start: appointmentRequestEvent.start,
            end: moment(appointmentRequestEvent.start).add(2, 'h'),
            allDay: appointmentRequestEvent.allDay,
            appointmentRequest: appointmentRequestEvent.appointmentRequest,
        };
        this.props.history.push(`${this.props.match.path}/set`, {
            appointmentRequestDropped: appointmentRequestDropped,
        });
    };

    updateAfterMoveOrResizeEvent(event) {
        let events = this.state.appointments;

        let appointment = event.appointment;
        appointment.startDateAndTime = moment(event.start).format();
        appointment.endDateAndTime = moment(event.end).format();

        const idx = events.indexOf(event);

        const updatedEvent = {...event, appointment: appointment};

        const nextEvents = [...events];
        nextEvents.splice(idx, 1, updatedEvent);

        this.updateAppointment(appointment, nextEvents);
    }

    updateAppointment = (appointment, nextEvents) => {
        let event = {
            appointmentId: appointment.appointmentId,
            date: moment(appointment.startDateAndTime).format('YYYY-MM-DD'),
            startTime: moment(appointment.startDateAndTime).format('HH:mm'),
            endTime: moment(appointment.endDateAndTime).format('HH:mm'),
            remarks: appointment.remarks,
            subject: appointment.subject,
            clientId: appointment.clientId,
        };

        appointmentsStorage.updateAppointment(event, this.serviceProviderHeaders)
            .then((response) => {
                this.setState({
                    appointments: nextEvents,
                })
            })
    };

    render() {
        const {appointments, page, totalPages} = this.state;
        const startIndex = page * TOTAL_PER_PAGE;

        return (
            <div>
                {/*<Page children={appointments} title={strings.mainPageStrings.APPOINTMENTS_PAGE_TITLE} columns={1}>*/}
                <div>
                    <Helmet>
                        <title>Meshekle | Appointments</title>
                    </Helmet>
                    <Grid stretched padded>
                        <Grid.Row>
                            <Header as="h1"
                                    floated="right">{strings.mainPageStrings.APPOINTMENTS_PAGE_TITLE}</Header>
                        </Grid.Row>
                        <Grid.Row columns='equal'>
                            <Grid.Column>
                                <Header as={'h3'} style={{'display': 'contents'}}> בקשות תורים:</Header>
                                {this.state.appointmentRequests.length === 0 ?
                                    <Header as={'h4'}> אין לך בקשות לתורים </Header>
                                    :
                                    <DraggableAppointmentRequest
                                        appointmentRequests={this.state.appointmentRequests}
                                        hoverOnAppointmentRequest={(appointmentRequest) => this.hoverOnAppointmentRequest.bind(this, appointmentRequest)}
                                        hoverOffAppointmentRequest={(appointmentRequest) => this.hoverOffAppointmentRequest.bind(this, appointmentRequest)}
                                        onClick={(appointmentRequest) => this.props.history.push(`${this.props.match.path}/requests/${appointmentRequest.requestId}`, {
                                            appointmentRequest: appointmentRequest
                                        })}
                                    />
                                }
                            </Grid.Column>
                            <Grid.Column width={13}>
                                <div style={{height: 500}}>
                                    <AppointmentCalendar
                                        events={Array.isArray(appointments) ? appointments : []}

                                        onSelectEvent={this.onSelectEvent.bind(this)}
                                        onSelectSlot={this.onSelectSlot.bind(this)}
                                        onDropAppointmentRequest={this.onDropAppointmentRequest.bind(this)}
                                        approveAppointmentRequest={(appointmentRequestDropped) => this.approveAppointmentRequest.bind(this, appointmentRequestDropped)}
                                        updateAfterMoveOrResizeEvent={this.updateAfterMoveOrResizeEvent}
                                    />
                                </div>
                            </Grid.Column>
                            {/*<Grid.Column>

                            </Grid.Column>*/}
                        </Grid.Row>
                    </Grid>
                </div>


                {/*<Table celled striped textAlign='right' selectable sortable>
                        <Table.Header>
                            <Table.Row>
                                <Table.HeaderCell>{strings.appointmentsPageStrings.CLIENT_ID}</Table.HeaderCell>
                                <Table.HeaderCell>{strings.appointmentsPageStrings.SERVICE_PROVIDER_ID}</Table.HeaderCell>
                                <Table.HeaderCell>{strings.appointmentsPageStrings.ROLE}</Table.HeaderCell>
                                <Table.HeaderCell>{strings.appointmentsPageStrings.SUBJECT}</Table.HeaderCell>
                                <Table.HeaderCell>{strings.appointmentsPageStrings.STATUS}</Table.HeaderCell>
                                <Table.HeaderCell>{strings.appointmentsPageStrings.DATE}</Table.HeaderCell>
                                <Table.HeaderCell>{strings.appointmentsPageStrings.START_TIME}</Table.HeaderCell>
                                <Table.HeaderCell>{strings.appointmentsPageStrings.END_TIME}</Table.HeaderCell>
                                <Table.HeaderCell>{strings.appointmentsPageStrings.REMARKS}</Table.HeaderCell>
                            </Table.Row>
                        </Table.Header>
                        <Table.Body>
                            {appointments.slice(startIndex, startIndex + TOTAL_PER_PAGE).map(appointment =>
                                (<Table.Row key={appointment.appointmentId}
                                            positive={this.state.highlightTableRow === appointment.appointmentId}>
                                    <Table.Cell>{appointment.AppointmentDetail.clientId}</Table.Cell>
                                    <Table.Cell>{appointment.AppointmentDetail.serviceProviderId}</Table.Cell>
                                    <Table.Cell>{appointment.AppointmentDetail.role}</Table.Cell>
                                    <Table.Cell>{appointment.AppointmentDetail.subject}</Table.Cell>
                                    <Table.Cell>{appointment.status}</Table.Cell>
                                    <Table.Cell>{new Date(appointment.startDateAndTime).toISOString().split('T')[0]}</Table.Cell>
                                    <Table.Cell>{new Date(appointment.startDateAndTime).toISOString().split('T')[1].split('.')[0].slice(0, -3)}</Table.Cell>
                                    <Table.Cell>{new Date(appointment.endDateAndTime).toISOString().split('T')[1].split('.')[0].slice(0, -3)}</Table.Cell>
                                    <Table.Cell>{appointment.remarks}</Table.Cell>
                                </Table.Row>),
                            )}
                        </Table.Body>
                        <Table.Footer>
                            <Table.Row>
                                <Table.HeaderCell colSpan={9}>
                                    <Menu floated="left" pagination>
                                        {page !== 0 && <Menu.Item as="a" icon onClick={this.decrementPage}>
                                            <Icon name="right chevron"/>
                                        </Menu.Item>}
                                        {times(totalPages, n =>
                                            (<Menu.Item as="a" key={n} active={n === page} onClick={this.setPage(n)}>
                                                {n + 1}
                                            </Menu.Item>),
                                        )}
                                        {page !== (totalPages - 1) &&
                                        <Menu.Item as="a" icon onClick={this.incrementPage}>
                                            <Icon name="left chevron"/>
                                        </Menu.Item>}
                                    </Menu>
                                </Table.HeaderCell>
                            </Table.Row>
                        </Table.Footer>
                    </Table>

                    <Button as={Link} to={`${this.props.match.path}/set`}
                            positive>{strings.appointmentsPageStrings.ADD_APPOINTMENT}</Button>
*/}
                {/*</Page>*/}


                <div>
                    <Switch>
                        <Route exec path={`${this.props.match.path}/requests/:appointmentRequestId`}
                               component={AppointmentRequestInfo}/>
                        <Route exec path={`${this.props.match.path}/set`} render={(props) => (
                            <AppointmentAdd {...props}
                                            approveAppointmentRequest={(appointmentRequestEvent) => this.approveAppointmentRequest(appointmentRequestEvent)}/>
                        )}/>
                        <Route exec path={`${this.props.match.path}/:appointmentId`}
                               component={AppointmentInfo}/>
                        <Route exec path={`${this.props.match.path}/:appointmentId/edit`}
                               component={AppointmentEdit}/>

                        {/*{
                            (this.props.location.pathname === window.location.pathname) ?
                            <Redirect to={`${this.props.match.path}`}/>
                            : null}
                        }*/}
                    </Switch>
                </div>
            </div>
            // </Grid>*/
        );
    }
}

export default AppointmentsManagementPage

