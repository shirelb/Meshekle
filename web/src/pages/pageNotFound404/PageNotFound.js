import React from 'react';
import {Grid, Header, Image} from 'semantic-ui-react';

const PageNotFound = () => (
    <Grid stretched textAlign={"center"}>
        <Grid.Column stretched textAlign={"center"}>
            <Image centered src={require('../../images/logo1.png')} size={'large'}/>
            <Header size={"huge"} textAlign={"center"} icon style={{marginTop: "-10px"}}>
                משקל'ה
                <Header.Subheader>יש משק ויש משקל'ה</Header.Subheader>
            </Header>

            <div style={{marginTop: 20, textAlign: "center"}}>
                <Header as={'h1'}> 404 </Header>
                <Header as={'h2'}> Page not found </Header>
            </div>
        </Grid.Column>
    </Grid>
);

export default PageNotFound;