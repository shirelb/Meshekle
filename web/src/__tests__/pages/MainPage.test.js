import React from 'react';


import {Redirect} from 'react-router-dom';
import {mount, shallow} from "enzyme/build";
import PageNotFound from "../../pages/pageNotFound404/PageNotFound";
import isLoggedIn from "../../shared/isLoggedIn";
import strings from "../../shared/strings";
import {setupComponent} from "../testHelpers";
import MainPage from "../../pages/mainPage/MainPage";
import {PhoneBookManagementPage} from "../../pages/phoneBookManagementPage/PhoneBookManagementPage";
import {AnnouncementsManagementPage} from "../../pages/announcementsManagementPage/AnnouncementsManagementPage";
import AppointmentsReportPage from "../../pages/appointmentsManagementPage/AppointmentsReportPage";
import AppointmentsManagementPage from "../../pages/appointmentsManagementPage/AppointmentsManagementPage";
import {ChoresManagementPage} from "../../pages/choresManagementPage/ChoresManagementPage";
import store from 'store';
import serviceProvidersStorage from "../../storage/serviceProvidersStorage";
import usersStorage from "../../storage/usersStorage";
import users from "../jsons/users";
import serviceProviders from "../jsons/serviceProviders";
import appointmentsStorage from "../../storage/appointmentsStorage";
import appointmentsOf549963652 from "../jsons/appointmentsServiceProvider549963652";
import appointmentRequestsOf549963652 from "../jsons/appointmentRequestsServiceProvider549963652";
import choresStorage from "../../storage/choresStorage";
import announcementsStorage from "../../storage/announcementsStorage";
import LoginPage from "../../pages/loginPage/LoginPage";


jest.mock("store");
jest.mock("../../shared/isLoggedIn");
jest.mock("../../storage/usersStorage");
jest.mock("../../storage/serviceProvidersStorage");
jest.mock("../../storage/appointmentsStorage");
jest.mock("../../storage/choresStorage");
jest.mock("../../storage/announcementsStorage");


describe("MainPage should", () => {
    let wrapper = null;
    let componentInstance = null;
    const props = {
        location: {
            pathname: '/home'
        },
        match: {
            isExact: true,
            path: '/home',
            url: '/home',
        }
    };
    let mockStore = {
        serviceProviderToken: null,
        serviceProviderId: "549963652",
        userId: "549963652",
    };

    const buildProps = (path) => {
        return {
            location: {
                pathname: path
            },
            match: {
                isExact: true,
                path: path,
                url: path,
            }
        };
    };

    store.get.mockImplementation((key) => mockStore[key]);
    store.remove.mockImplementation((key) => mockStore[key] = null);
    usersStorage.getUsers.mockResolvedValue(users);
    usersStorage.getUserByUserID.mockImplementation((userId) => Promise.resolve(users.filter(user => user.userId === userId)[0]));
    serviceProvidersStorage.getServiceProviderUserDetails.mockImplementation((serviceProviderId) => Promise.resolve({data: users.filter(user => user.userId === serviceProviderId)[0]}));
    serviceProvidersStorage.getServiceProviderById.mockImplementation((serviceProviderId) => Promise.resolve(serviceProviders.filter(provider =>
        provider.serviceProviderId === serviceProviderId)));
    serviceProvidersStorage.getServiceProviderPermissionsById.mockResolvedValue(["appointments"]);
    serviceProvidersStorage.getRolesOfServiceProvider.mockResolvedValue(["appointmentsHairDresser"]);
    serviceProvidersStorage.getServiceProviders.mockResolvedValue(serviceProviders);
    appointmentsStorage.getServiceProviderAppointments.mockResolvedValue({data: appointmentsOf549963652});
    appointmentsStorage.getServiceProviderAppointmentRequests.mockImplementation(() => Promise.resolve({data: JSON.parse(JSON.stringify(appointmentRequestsOf549963652))}));
    choresStorage.getAllChoreTypes.mockResolvedValue([]);
    choresStorage.getChoreTypeSetting.mockResolvedValue({data: []});
    choresStorage.getUserChoresForType.mockResolvedValue({data: []});
    choresStorage.getUsersForChoreType.mockResolvedValue({data: []});
    choresStorage.getReplacementRequests.mockResolvedValue({data: []});
    announcementsStorage.getUsers.mockResolvedValue({data: users});
    announcementsStorage.getCategories.mockResolvedValue({data: [{categoryId:1,serviceProviderId:"544359761",categoryName:"ביטחון"}]});
    announcementsStorage.getCategoriesByServiceProviderId.mockResolvedValue({data: []});
    announcementsStorage.getAnnouncements.mockResolvedValue({data: []});
    announcementsStorage.getAnnouncementsRequests.mockResolvedValue({data: []});

    beforeAll(async () => {

    });

    beforeEach(() => {
        jest.clearAllMocks();
        mockStore['serviceProviderToken'] = "SOME TOKEN eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9";
    });

    afterEach(() => {
    });

    test("render Login when user is NOT authenticated", async () => {
        mockStore['serviceProviderToken'] = null;

        isLoggedIn.mockResolvedValue(false);
        const arrResponse = await setupComponent('mount', MainPage, null, props, "/login");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        wrapper.update();

        expect(wrapper.find(Redirect)).toHaveLength(1);
        expect(wrapper.find('Router').prop('history').location.pathname).toEqual('/login');
    });

    test("render Home when user is authenticated", async () => {
        mockStore['serviceProviderToken'] = "SOME TOKEN eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9";

        isLoggedIn.mockResolvedValue(true);
        const arrResponse = await setupComponent('shallow', MainPage, null, props, "/home");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        expect(componentInstance).toMatchSnapshot();

        expect(wrapper.find(MainPage)).toHaveLength(1);
        expect(wrapper.find(Redirect)).toHaveLength(0);
        expect(wrapper.find(PageNotFound)).toHaveLength(0);
    });

    test("render a page with header and form with two fields and a send button", async () => {
        isLoggedIn.mockResolvedValue(true);
        const arrResponse = await setupComponent('mount', MainPage, null, props, "/home");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];
        const mountWrapper = wrapper.find('MainPage');

        expect(mountWrapper.find('Sidebar')).toHaveLength(1);
        expect(mountWrapper.find('MenuItem')).toHaveLength(4);
        expect(mountWrapper.find('MenuItem').at(0).props().name).toEqual('home');
        expect(mountWrapper.find('MenuItem').at(0).props().children[1]).toEqual(strings.mainPageStrings.MAIN_PAGE_TITLE);
        expect(mountWrapper.find('MenuItem').at(1).props().name).toEqual('phoneBook');
        expect(mountWrapper.find('MenuItem').at(1).props().children[1]).toEqual(strings.mainPageStrings.PHONE_BOOK_PAGE_TITLE);
        expect(mountWrapper.find('MenuItem').at(2).props().name).toEqual('appointments');
        expect(mountWrapper.find('MenuItem').at(2).props().children[1]).toEqual(strings.mainPageStrings.APPOINTMENTS_PAGE_TITLE);
        expect(mountWrapper.find('MenuItem').at(3).props().name).toEqual('logout');
        expect(mountWrapper.find('MenuItem').at(3).props().children[1]).toEqual(strings.mainPageStrings.LOGOUT);

        expect(mountWrapper.find('Home')).toHaveLength(1);
        expect(mountWrapper.find('Home').find('Grid')).toHaveLength(2);
        expect(mountWrapper.find('Home').find('Image')).toHaveLength(1);
        expect(mountWrapper.find('Home').find('Image').props('src').src).toEqual('logo1.png');
        expect(mountWrapper.find('Home').find('Header')).toHaveLength(5);
        expect(mountWrapper.find('Home').find('Header').at(0).props().children[0]).toEqual('משקל\'ה');
        expect(mountWrapper.find('Home').find('Header').find('HeaderSubheader').props().children).toEqual('יש משק ויש משקל\'ה');
        expect(mountWrapper.find('Home').find('Header').at(1).props().children).toEqual('ברוכים הבאים');
        expect(mountWrapper.find('Home').find('Header').at(2).props().children).toEqual('Padget Creaser');
        expect(mountWrapper.find('Home').find('Header').at(3).find('GridRow').at(0).props().children).toEqual('התפקידים באחריותך הם:');
        expect(mountWrapper.find('Home').find('Header').at(3).find('GridRow').at(1).props().children).toHaveLength(1);
        expect(mountWrapper.find('Home').find('Header').at(4).props().children[1]).toEqual('מספרה');
        expect(mountWrapper.find('Home').find('Header').at(4).find('Icon').props().name).toEqual('calendar alternate outline');
    });

    test("render Home on /home", async () => {
        isLoggedIn.mockResolvedValue(true);
        const arrResponse = await setupComponent('mount', MainPage, null, props, "/home");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        expect(wrapper.find('Home')).toHaveLength(1);
    });

    test("render PhoneBookManagementPage for /phoneBook", async () => {
        isLoggedIn.mockResolvedValue(true);
        const arrResponse = await setupComponent('mount', MainPage, null, buildProps("/phoneBook"), "/phoneBook");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        expect(wrapper.find(PhoneBookManagementPage)).toHaveLength(1);
    });

    test("render AnnouncementsManagementPage for /announcements", async () => {
        isLoggedIn.mockResolvedValue(true);
        const arrResponse = await setupComponent('mount', MainPage, null, buildProps("/announcements"), "/announcements");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        expect(wrapper.find(AnnouncementsManagementPage)).toHaveLength(1);
    });

    test("render AppointmentsReportPage for /appointments/report", async () => {
        isLoggedIn.mockResolvedValue(true);
        const arrResponse = await setupComponent('mount', MainPage, null, buildProps("/appointments/report"), "/appointments/report");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        expect(wrapper.find(AppointmentsReportPage)).toHaveLength(1);
    });

    test("render AppointmentsManagementPage for /appointments", async () => {
        isLoggedIn.mockResolvedValue(true);
        const arrResponse = await setupComponent('mount', MainPage, null, buildProps("/appointments"), "/appointments");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        expect(wrapper.find(AppointmentsManagementPage)).toHaveLength(1);
    });

    test("render ChoresManagementPage for /chores", async () => {
        isLoggedIn.mockResolvedValue(true);
        const arrResponse = await setupComponent('mount', MainPage, null, buildProps("/chores"), "/chores");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        expect(wrapper.find(ChoresManagementPage)).toHaveLength(1);
    });

    test("render PageNotFound for /random", async () => {
        isLoggedIn.mockResolvedValue(true);
        const arrResponse = await setupComponent('mount', MainPage, null, buildProps("/random"), "/random");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        expect(wrapper.find(PageNotFound)).toHaveLength(1);
    });

    test("open Home on click Home button on Sidebar Menu", async () => {
        isLoggedIn.mockResolvedValue(true);
        const arrResponse = await setupComponent('mount', MainPage, history, buildProps("/appointments"), "/appointments");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        // const settingsBtn = wrapper.find('MainPage').find('NavLink').first();
        const menuBtn = wrapper.find('MainPage').find('MenuItem').at(0);
        await menuBtn.simulate('click', {button: 0});

        expect(wrapper.find('Router').prop('history').location.pathname).toEqual('/home');
        expect(wrapper.find('Home')).toHaveLength(1);
    });

    test("open PhoneBook on click Phonebook button on Sidebar Menu", async () => {
        isLoggedIn.mockResolvedValue(true);
        const arrResponse = await setupComponent('mount', MainPage, history, props, "/home");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        const menuBtn = wrapper.find('MainPage').find('MenuItem').at(1);
        await menuBtn.simulate('click', {button: 0});

        expect(wrapper.find('Router').prop('history').location.pathname).toEqual('/phoneBook');
        expect(wrapper.find(PhoneBookManagementPage)).toHaveLength(1);
    });

    test("open Appointments on click Appointments button on Sidebar Menu", async () => {
        isLoggedIn.mockResolvedValue(true);
        const arrResponse = await setupComponent('mount', MainPage, history, props, "/home");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        const menuBtn = wrapper.find('MainPage').find('MenuItem').at(2);
        await menuBtn.simulate('click', {button: 0});

        expect(wrapper.find('Router').prop('history').location.pathname).toEqual('/appointments');
        expect(wrapper.find(AppointmentsManagementPage)).toHaveLength(1);
    });

    test("logout on click Logout button on Sidebar Menu", async () => {
        isLoggedIn.mockResolvedValue(true);
        const arrResponse = await setupComponent('mount', MainPage, history, props, "/home");
        wrapper = arrResponse[0];
        componentInstance = arrResponse[1];

        const menuBtn = wrapper.find('MainPage').find('MenuItem').at(3);
        await menuBtn.simulate('click', {button: 0});

        expect(wrapper.find(Redirect)).toHaveLength(1);
    });

});
