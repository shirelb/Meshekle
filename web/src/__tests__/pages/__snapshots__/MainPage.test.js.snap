// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage should render Home when user is authenticated 1`] = `
MainPage {
  "context": Object {},
  "getServiceProviderHomeDetails": [Function],
  "handleLogout": [Function],
  "hasPhoneBookPermissions": [Function],
  "props": Object {
    "location": Object {
      "pathname": "/home",
    },
    "match": Object {
      "isExact": true,
      "path": "/home",
      "url": "/home",
    },
  },
  "refs": Object {},
  "serviceProviderHeaders": Object {
    "Authorization": "Bearer SOME TOKEN eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9",
  },
  "setState": [Function],
  "state": Object {
    "serviceProviderPermissions": Array [
      "appointments",
    ],
    "serviceProviderRoles": Array [
      "appointmentsHairDresser",
    ],
    "userFullname": "Padget Creaser",
  },
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {},
      "_didScheduleRenderPhaseUpdate": false,
      "_dispatcher": Object {
        "readContext": [Function],
        "useCallback": [Function],
        "useContext": [Function],
        "useDebugValue": [Function],
        "useEffect": [Function],
        "useImperativeHandle": [Function],
        "useLayoutEffect": [Function],
        "useMemo": [Function],
        "useReducer": [Function],
        "useRef": [Function],
        "useState": [Function],
      },
      "_element": <MainPage
        location={
          Object {
            "pathname": "/home",
          }
        }
        match={
          Object {
            "isExact": true,
            "path": "/home",
            "url": "/home",
          }
        }
      />,
      "_firstWorkInProgressHook": null,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_isReRender": false,
      "_newState": null,
      "_numberOfReRenders": 0,
      "_renderPhaseUpdates": null,
      "_rendered": <div>
        <HelmetWrapper
          defer={true}
          encodeSpecialCharacters={true}
        >
          <title>
            Meshekle
          </title>
        </HelmetWrapper>
        <Sidebar
          as={[Function]}
          direction="right"
          duration={500}
          icon="labeled"
          inverted={true}
          vertical={true}
          visible={true}
          width="thin"
        >
          <MenuItem
            as={[Function]}
            name="home"
            to="/home"
          >
            <Icon
              as="i"
              name="home"
            />
            בית
          </MenuItem>
          <MenuItem
            as={[Function]}
            name="phoneBook"
            to="/phoneBook"
          >
            <Icon
              as="i"
              name="address book outline"
            />
            ספר טלפונים
          </MenuItem>
          <MenuItem
            as={[Function]}
            name="appointments"
            to="/appointments"
          >
            <Icon
              as="i"
              name="calendar alternate outline"
            />
            תורים
          </MenuItem>
          <MenuItem
            name="logout"
            onClick={[Function]}
          >
            <Icon
              as="i"
              name="power"
            />
            התנתק
          </MenuItem>
        </Sidebar>
        <div
          className="mainBody"
        >
          <Switch>
            <Route
              exec={true}
              path="/home"
              render={[Function]}
            />
            <Route
              exec={true}
              path="/phoneBook"
              render={[Function]}
            />
            <Route
              component={[Function]}
              exec={true}
              path="/appointments/report"
            />
            <Route
              component={[Function]}
              path="/appointments"
            />
            <Route
              component={[Function]}
              path="/chores"
            />
            <Route
              component={[Function]}
              path="/announcements"
            />
            <Route
              component={[Function]}
            />
          </Switch>
        </div>
      </div>,
      "_rendering": false,
      "_updater": [Circular],
      "_workInProgressHook": null,
    },
  },
}
`;
